### PR TITLE
Pfs 50 invoices backend

### PR DESCRIPTION
**Pull request by Nicholassully**

### CODE REVIEW

File: cypress/e2e/invoices.cy.js [hunk 1]
```diff
@@ -32,24 +30,23 @@ describe("Invoices ledger allocations", () => {
...
+//         cy.contains('[data-cy="supervision-level"]', "Supervision level");
+//         cy.contains('[data-cy="supervision-amount"]', "Amount");
+//         cy.contains('[data-cy="supervision-from"]', "From");
+//         cy.contains('[data-cy="supervision-to"]', "To");
+//         cy.contains('[data-cy="supervision-level-data"]', "General").first();
+//         cy.contains('[data-cy="supervision-amount-data"]', "£320").first();
+//         cy.get('[data-cy="supervision-from-data"]').contains("01/04/2019").first();
+//         cy.get('[data-cy="supervision-to-data"]').contains("31/03/2020").first();
+//     });
+// });
```
Comment (josephsmith0705) [inline cypress/e2e/invoices.cy.js:52]: Will we be able to uncomment these for some future work? Or, can we remove these commented lines

**Review by josephsmith0705 — DISMISSED** [review]

File: cypress/e2e/invoices.cy.js (hunk 1 above)
Comment (Nicholassully) [reply cypress/e2e/invoices.cy.js:52]: Yes in the next ticket this test will be needed

Comment (denialanderror) [reply cypress/e2e/invoices.cy.js:52]: Could you stick a TODO comment with the ticket number it is meant to be done on, so if it gets missed we know when it should have been?

**Review by denialanderror — CHANGES_REQUESTED** [review]

File: cypress/e2e/invoices.cy.js (hunk 1 above)
Comment (Nicholassully) [reply cypress/e2e/invoices.cy.js:52]: done

File: finance-api/cmd/api/invoices.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	"encoding/json"
+	"net/http"
+	"strconv"
+)
+
+func (s *Server) getInvoices(w http.ResponseWriter, r *http.Request) {
+	clientId, _ := strconv.Atoi(r.PathValue("id"))
+	accountInfo, err := s.Service.GetInvoices(clientId)
+
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply finance-api/cmd/api/invoices.go:13]: done

File: finance-api/internal/service/invoices.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+import (
+	"context"
+	"github.com/jackc/pgx/v5/pgtype"
+	"github.com/opg-sirius-finance-hub/shared"
+	"regexp"
+	"strconv"
+)
+
+func (s *Service) GetInvoices(id int) (*shared.Invoices, error) {
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply finance-api/internal/service/invoices.go:11]: done

File: finance-api/internal/service/invoices.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+	fc, err := s.Store.GetInvoices(ctx, pgtype.Int4{Int32: int32(id), Valid: true})
+	if err != nil {
+		return nil, err
+	}
+
+	ledgerAllocationsData, err := s.Store.GetLedgerAllocations(ctx, pgtype.Int4{Int32: int32(id), Valid: true})
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply finance-api/internal/service/invoices.go:22]: done

File: finance-api/internal/service/invoices.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/jackc/pgx/v5/pgtype"
+	"github.com/opg-sirius-finance-hub/shared"
+	"regexp"
+	"strconv"
+)
+
+func (s *Service) GetInvoices(id int) (*shared.Invoices, error) {
+	ctx := context.Background()
+
+	fc, err := s.Store.GetInvoices(ctx, pgtype.Int4{Int32: int32(id), Valid: true})
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply finance-api/internal/service/invoices.go:14]: Yeah I was trying not to do anything like what I did but the sql thats generated makes the data type pgtype.Int4 so I do need to change the type in less there is a better way

File: finance-api/internal/service/invoices.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+	var ledgerAllocations []shared.Ledger
+	var invoices shared.Invoices
+	var totalOfLedgerAllocationsAmount int
+
+	for _, i := range ledgerAllocationsData {
+		var ledgerAllocation = shared.Ledger{
+			Amount:          IntToDecimalString(int(i.Amount)),
+			ReceivedDate:    shared.Date{Time: i.Allocateddate.Time},
+			TransactionType: "unknown",
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply finance-api/internal/service/invoices.go:32]: Something we need to go through together on this one

File: finance-api/internal/service/invoices.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	var ledgerAllocations []shared.Ledger
+	var invoices shared.Invoices
+	var totalOfLedgerAllocationsAmount int
+
+	for _, i := range ledgerAllocationsData {
+		var ledgerAllocation = shared.Ledger{
+			Amount:          IntToDecimalString(int(i.Amount)),
+			ReceivedDate:    shared.Date{Time: i.Allocateddate.Time},
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply finance-api/internal/service/invoices.go:31]: Then we cant use the shared struct I thought?

File: finance-api/internal/store/queries/finance_client.sql [hunk 8]
```diff
@@ -1,2 +1,8 @@
 -- name: GetAccountInformation :one
-SELECT cacheddebtamount, cachedcreditamount, payment_method FROM finance_client WHERE client_id = $1;
\ No newline at end of file
+SELECT cacheddebtamount, cachedcreditamount, payment_method FROM finance_client WHERE client_id = $1;
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply finance-api/internal/store/queries/finance_client.sql:2]: done

File: finance-api/internal/store/queries/finance_client.sql [hunk 9]
```diff
@@ -1,2 +1,8 @@
...
+
+-- name: GetInvoices :many
+SELECT id, reference, amount, raiseddate, cacheddebtamount FROM invoice WHERE finance_client_id = $1;
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply finance-api/internal/store/queries/finance_client.sql:5]: The invoice will be in this pr yes

File: finance-hub/internal/api/get_invoices_test.go [hunk 10]
```diff
@@ -94,17 +94,16 @@ func TestGetInvoicesCanThrow500Error(t *testing.T) {
...
 
-	clientList, err := client.GetInvoices(getContext(nil), 3)
-
-	var expectedResponse shared.Invoices
-
-	assert.Equal(t, expectedResponse, clientList)
+	_, err := client.GetInvoices(getContext(nil), 1)
+	//var expectedResponse shared.Invoices
+	//
+	//assert.Equal(t, expectedResponse, invoices)
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply finance-hub/internal/api/get_invoices_test.go:102]: done

File: finance-api/internal/service/invoices.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (denialanderror) [reply finance-api/internal/service/invoices.go:45]: The shared struct is the contract between the front end and the back end, not the representation presented to the user. If we need to display data in a different way to how it is fetched, you will need to transform it into a different struct (see how it is done in `finance-hub/internal/server/fee_reductions.go`)

File: finance-api/internal/service/invoices.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+	ctx := context.Background()
+	group, _ := errgroup.WithContext(ctx)
+
+	var invoicesData []store.GetInvoicesRow
+	var ledgerAllocationsData []store.GetLedgerAllocationsRow
+	var ledgerAllocations []shared.Ledger
+	var invoices shared.Invoices
+	var totalOfLedgerAllocationsAmount int
+
+	group.Go(func() error {
```
Comment (denialanderror) [inline finance-api/internal/service/invoices.go:21]: On second thought, why are you making two calls to the database here? If we want to fetch invoices along with their ledger allocations, can't we do it as a single query? We no longer have an ORM in charge of our queries.

**Review by josephsmith0705 — APPROVED** [review]

**Pull request merged**